### PR TITLE
Custom inaugurator image support

### PR DESCRIPTION
**Pull request by michaelk-stratoscale**
@eliran-stratoscale @rom-stratoscale @adir-stratoscale

### CODE REVIEW

File: rackattack/common/tftpboot.py [hunk 1]
```diff
@@ -82,10 +85,35 @@ def _configurationForInaugurator(self, id, mac, ip, serialPort, clearDisk, targe
...
+                return os.path.join(os.path.dirname(INAUGURATOR_INITRD), image)
+        return INAUGURATOR_INITRD
+
+    def _getKernelImage(self, id):
+        for pattern, image in self._imagesMap['kernel'].iteritems():
+            if re.search(pattern, id):
+                return os.path.join(os.path.dirname(INAUGURATOR_INITRD), image)
+        return INAUGURATOR_KERNEL
+
+    def loadImageMaps(self):
```
Comment (adir-stratoscale) [inline rackattack/common/tftpboot.py]: why not private?

**Comment by adir-stratoscale** [comment]
@michaelk-stratoscale can you describe how the mapping server_id-inaugurator is defined?

File: rackattack/common/tftpboot.py [hunk 2]
```diff
@@ -82,10 +85,35 @@ def _configurationForInaugurator(self, id, mac, ip, serialPort, clearDisk, targe
...
                                                                targetDevice, targetDeviceType),
-            inauguratorKernel=os.path.basename(INAUGURATOR_KERNEL),
-            inauguratorInitrd=os.path.basename(INAUGURATOR_INITRD),
+            inauguratorKernel=self._getKernelImage(id),
+            inauguratorInitrd=self._getInauguratorImage(id),
             serialPort=serialPort)
 
+    def _getInauguratorImage(self, id):
+        for pattern, image in self._imagesMap['inaugurator'].iteritems():
+            if re.search(pattern, id):
```
Comment (adir-stratoscale) [inline rackattack/common/tftpboot.py]: what if there are multiple entries that are matched the given id?

File: rackattack/common/tftpboot.py [hunk 3]
```diff
@@ -82,10 +85,35 @@ def _configurationForInaugurator(self, id, mac, ip, serialPort, clearDisk, targe
...
+                return os.path.join(os.path.dirname(INAUGURATOR_INITRD), image)
+        return INAUGURATOR_INITRD
+
+    def _getKernelImage(self, id):
+        for pattern, image in self._imagesMap['kernel'].iteritems():
+            if re.search(pattern, id):
```
Comment (adir-stratoscale) [inline rackattack/common/tftpboot.py]: same here.

File: rackattack/common/tftpboot.py (hunk 2 above)
Comment (adir-stratoscale) [inline rackattack/common/tftpboot.py]: also, note that dict has no order, so different runs have different results.

File: rackattack/common/tftpboot.py [hunk 4]
```diff
@@ -82,10 +85,35 @@ def _configurationForInaugurator(self, id, mac, ip, serialPort, clearDisk, targe
...
+                return os.path.join(os.path.dirname(INAUGURATOR_INITRD), image)
+        return INAUGURATOR_KERNEL
+
+    def loadImageMaps(self):
+        self._imagesMap = {"inaugurator": dict(), "kernel": dict()}
+        for imageType, imageInstance in [('inaugurator', INAUGURATOR_INITRD), ('kernel', INAUGURATOR_KERNEL)]:
+            _imageList = glob.glob("%s/*.%s" % (os.path.dirname(ROOT_PATH), os.path.basename(imageInstance)))
+            for instance in _imageList:
+                prefix = os.path.basename(instance).split(".")[0]
```
Comment (eliran-stratoscale) [inline rackattack/common/tftpboot.py]: Use rsplit with max=1

File: rackattack/common/tftpboot.py [hunk 5]
```diff
@@ -82,10 +85,35 @@ def _configurationForInaugurator(self, id, mac, ip, serialPort, clearDisk, targe
...
+
+    def _getKernelImage(self, id):
+        for pattern, image in self._imagesMap['kernel'].iteritems():
+            if re.search(pattern, id):
+                return os.path.join(os.path.dirname(INAUGURATOR_INITRD), image)
+        return INAUGURATOR_KERNEL
+
+    def loadImageMaps(self):
+        self._imagesMap = {"inaugurator": dict(), "kernel": dict()}
+        for imageType, imageInstance in [('inaugurator', INAUGURATOR_INITRD), ('kernel', INAUGURATOR_KERNEL)]:
```
Comment (eliran-stratoscale) [inline rackattack/common/tftpboot.py]: Don't use a loop. Instead wrap this in a method and call it twice.

File: rackattack/common/tftpboot.py [hunk 6]
```diff
@@ -82,10 +85,35 @@ def _configurationForInaugurator(self, id, mac, ip, serialPort, clearDisk, targe
...
+            _imageList = glob.glob("%s/*.%s" % (os.path.dirname(ROOT_PATH), os.path.basename(imageInstance)))
+            for instance in _imageList:
```
Comment (eliran-stratoscale) [inline rackattack/common/tftpboot.py]: instance -> imagePath

File: rackattack/common/tftpboot.py [hunk 7]
```diff
@@ -82,10 +85,35 @@ def _configurationForInaugurator(self, id, mac, ip, serialPort, clearDisk, targe
...
+                prefix = os.path.basename(instance).split(".")[0]
+                self._imagesMap[imageType][self._prefixToRegex(prefix)] = instance
```
Comment (eliran-stratoscale) [inline rackattack/common/tftpboot.py]: Seperate _imageMap to _kernelImageMap and _inauguratorImageMap

File: rackattack/common/tftpboot.py [hunk 8]
```diff
@@ -82,10 +85,35 @@ def _configurationForInaugurator(self, id, mac, ip, serialPort, clearDisk, targe
...
+
+    def _prefixToRegex(self, prefix):
+        for i in range(10, 0, -1):
+            prefix = prefix.replace("+" * i, "\d{%d}" % i)
```
Comment (eliran-stratoscale) [inline rackattack/common/tftpboot.py]: \d\d sounds simpler than \d{2}

**Comment by eliran-stratoscale** [comment]
Call this from `reloadConfiguration`.

File: rackattack/common/tftpboot.py (hunk 2 above)
Comment (michaelk-stratoscale) [inline rackattack/common/tftpboot.py]: @adir-stratoscale, I prefer providing additional flexibility on expense of coherency. In other words, user is responsible in for non-colliding image marking.

File: rackattack/common/tftpboot.py (hunk 3 above)
Comment (michaelk-stratoscale) [inline rackattack/common/tftpboot.py]: Same answer :)

File: rackattack/common/tftpboot.py (hunk 1 above)
Comment (michaelk-stratoscale) [inline rackattack/common/tftpboot.py]: This comment was relevant until we identified a gap, whereas this method will be called from another class to provide dynamic reloading of the configuration.

File: rackattack/common/tftpboot.py (hunk 4 above)
Comment (michaelk-stratoscale) [inline rackattack/common/tftpboot.py]: Done

File: rackattack/common/tftpboot.py (hunk 6 above)
Comment (michaelk-stratoscale) [inline rackattack/common/tftpboot.py]: Done

File: rackattack/common/tftpboot.py (hunk 5 above)
Comment (michaelk-stratoscale) [inline rackattack/common/tftpboot.py]: Done

File: rackattack/common/tftpboot.py (hunk 8 above)
Comment (michaelk-stratoscale) [inline rackattack/common/tftpboot.py]: Done

File: rackattack/common/tftpboot.py (hunk 7 above)
Comment (michaelk-stratoscale) [inline rackattack/common/tftpboot.py]: Done

**Pull request closed without merging**